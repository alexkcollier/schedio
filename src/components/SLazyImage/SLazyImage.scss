.lazy-image {
  transition: filter duration() easing('out'), -webkit-filter duration() easing('out');

  &--loading {
    background-color: color('grey', 'light');
    filter: blur(5px);

    // this is needed so Safari keeps sharp edges
    transform: scale(1);
  }
}
