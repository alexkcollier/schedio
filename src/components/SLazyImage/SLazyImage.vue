<template>
  <img
    ref="lazyElement"
    :src="imageToLoad"
    :srcset="deferredSrcset"
    :sizes="deferredSizes"
    class="lazy-image"
    :class="{
      'lazy-image--loading': !isLoaded,
    }"
    :style="implicitPlaceholder"
    @load="handleLoad"
  >
</template>

<script>
export default {
  name: 'SLazyImage',

  props: {
    src: {
      type: String,
      required: true,
    },

    placeholder: {
      type: String,
      default: '',
    },

    intersectionOptions: {
      type: Object,
      default: () => ({}),
    },

    placeholderHeight: {
      type: String,
      default: '100%',
    },

    width: {
      type: String,
      default: '100%',
    },
  },

  data () {
    return {
      observer: null,
      isIntersecting: false,
      isLoaded: false,
      usePadding: true,
    }
  },

  computed: {
    canLoad () {
      // skip lazy loading if `IntersectionObserver` not supported
      return !('IntersectionObserver' in window) || this.isIntersecting
    },

    implicitPlaceholder () {
      // prevents reflow while image is loading
      const shouldUsePadding = this.usePadding || !this.hasContent

      return {
        // padding handles when image is `display: inline`
        paddingTop: shouldUsePadding ? this.placeholderHeight : null,
        paddingLeft: shouldUsePadding ? this.width : null,
        width: this.hasContent ? this.width : 0,
      }
    },

    hasContent () {
      return !!this.isLoaded || !!this.placeholder
    },

    imageToLoad () {
      return this.deferUntilIntersection(this.src, this.placeholder)
    },

    deferredSrcset () {
      return this.deferUntilIntersection(this.$attrs.srcset)
    },

    deferredSizes () {
      return this.deferUntilIntersection(this.$attrs.sizes)
    },
  },

  mounted () {
    this.createObserver(this.$refs.lazyElement)
  },

  destroyed () {
    this.observer.disconnect()
  },

  methods: {
    handleLoad (event) {
      // using padding at first ensures content does not reflow
      this.usePadding = false
      // tell vm to de-blur the loaded image
      if (this.canLoad) this.isLoaded = true
    },

    deferUntilIntersection (prop, preIntersection = null) {
      return this.canLoad && prop ? prop : preIntersection
    },

    createObserver (el) {
      if ('IntersectionObserver' in window) {
        this.observer = new IntersectionObserver(this.onIntersection, this.intersectionOptions)
        this.observer.observe(el)
      }
    },

    onIntersection (entries, observer) {
      entries.forEach(({ isIntersecting, target }) => {
        if (isIntersecting) {
          // tell the image to load and stop observing
          this.isIntersecting = isIntersecting
          observer.unobserve(target)
        }
      })
    },
  },
}
</script>

<style lang="scss">
.lazy-image {
  transition: filter duration() easing('out'), -webkit-filter duration() easing('out');

  &--loading {
    background-color: color('grey', 'light');
    filter: blur(5px);

    // this is needed so Safari keeps sharp edges
    transform: scale(1);
  }
}
</style>
