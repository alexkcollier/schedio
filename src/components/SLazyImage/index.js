import SLazyImage from './SLazyImage.vue'

export { SLazyImage }

export default {
  install: Vue => Vue.component(SLazyImage.name, SLazyImage),
}
