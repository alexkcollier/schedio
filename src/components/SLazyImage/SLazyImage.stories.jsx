import { SLazyImage } from '@/components/SLazyImage';
import { SHeading } from '@/components/SHeading';
import { text, object } from '@storybook/addon-knobs';
import PropList from '@@/docs/components/PropList';

export default {
  title: 'Components/SLazyImage',
  component: SLazyImage,
  parameters: {
    jest: 'SLazyImage',
  },
};

const defaultImage = 'https://source.unsplash.com/a7XPK-oDsMk';
const intersectionOptions = {
  root: null,
  rootMargin: '0px',
  threshold: 0,
};

export const LazyImage = () => ({
  props: {
    props: {
      default: {
        src: text('src', defaultImage, 'Optional Props'),
        placeholder: text('placeholder', `${defaultImage}/16x25`, 'Optional Props'),
        placeholderHeight: text('placeholder-height', '150%', 'Optional Props'),
        width: text('width', '330px', 'Optional Props'),
        intersectionOptions: object('instersection-options', intersectionOptions, 'Optional Props'),
      },
    },
  },

  render (h) {
    const { props } = this.$props;

    return (
      <div>
        <SHeading>Lazy Image</SHeading>
        <p>
          <code>&lt;SLazyImage /&gt;</code> provides lazy loading for images. This component is
          based on the <code>IntersectionObserver</code> API.
        </p>
        <p>
          <code>placeholder-height</code> works best when set to a percentage representing the
          aspect ratio of the image.
        </p>

        <SHeading level="2">Example</SHeading>
        <p>There are two images 500px apart. Change any knobs now to see their effects.</p>
        <SLazyImage {...{ props }} />
        <div style="height: 500px;"></div>
        <SLazyImage {...{ props }} />

        {/* This doesn't work if props are only generated through a mixin */}
        {SLazyImage.props && <PropList component={SLazyImage} />}
      </div>
    );
  },
});

LazyImage.story = {
  name: 'Lazy Image',
};
