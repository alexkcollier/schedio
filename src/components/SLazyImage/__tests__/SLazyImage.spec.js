import { shallowMount } from '@vue/test-utils';
import SLazyImage from '@/components/SLazyImage/SLazyImage.vue';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);

describe('SLazyImage.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { });
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(SLazyImage, {
      propsData: {
        alt: 'alt',
        src: 'https://source.unsplash.com/a7XPK-oDsMk',
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('requires src', () => {
    shallowMount(SLazyImage);
    expect(errorSpy).toHaveBeenCalledTimes(1);
  });
});
