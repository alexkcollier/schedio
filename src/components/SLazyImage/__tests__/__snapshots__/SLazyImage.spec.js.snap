// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLazyImage.vue renders correctly 1`] = `<img src="https://source.unsplash.com/a7XPK-oDsMk" class="lazy-image lazy-image--loading" style="padding-top: 100%; padding-left: 100%; width: 0px;" alt="alt">`;
